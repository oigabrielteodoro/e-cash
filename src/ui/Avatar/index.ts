export * from './Avatar'
export * from './NotFoundAvatar'
