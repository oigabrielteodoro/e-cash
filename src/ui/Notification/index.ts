export const NOTIFICATION_EVENT_NAME = 'notification'
export const EXPIRES_NOTIFICATION_IN_SECONDS = 1000 * 3 // 3 seconds

export * from './Notification'
export * from './NotificationContainer'
