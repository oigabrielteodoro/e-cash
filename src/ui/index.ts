export * from './Style'
export * from './Input'
export * from './Button'
export * from './PasswordInput'
export * from './Toast'
export * from './Logo'
export * from './SideBar'
export * from './AppLayout'
export * from './Tooltip'
export * from './Avatar'
export * from './ShimmerEffect'
export * from './Popover'
export * from './Grid'
