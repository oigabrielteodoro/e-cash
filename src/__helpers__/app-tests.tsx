import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
