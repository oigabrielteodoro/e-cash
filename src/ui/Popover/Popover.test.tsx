import React from 'react'

import { Popover } from 'ui'
import { render, screen, userEvent, waitFor } from '__helpers__/app-tests'

function MockedComponent() {
  return (
    <Popover position='top' innerContent={<h1>Popover content</h1>}>
      Open popover
    </Popover>
  )
}

describe('Popover', () => {
  it('should be able render correctly', () => {
    render(<MockedComponent />)

    expect(
      screen.getByRole('button', {
        name: 'Open popover',
      }),
    ).toBeInTheDocument()
  })

  it('should be able open and close popover', async () => {
    render(<MockedComponent />)

    const button = screen.getByRole('button', {
      name: 'Open popover',
    })

    expect(button).toBeInTheDocument()

    userEvent.click(button)

    expect(await screen.findByText('Popover content')).toBeInTheDocument()

    userEvent.click(button)

    await waitFor(() =>
      expect(screen.queryByText('Popover content')).toBeInTheDocument(),
    )
  })
})
