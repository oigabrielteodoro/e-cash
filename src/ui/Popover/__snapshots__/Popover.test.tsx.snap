// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should be able match snapshot 1`] = `
.c0 {
  width: 100%;
  position: relative;
}

.c1 {
  width: 100%;
  border: 0;
  background: transparent;
  -webkit-transition: 300ms;
  transition: 300ms;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      Open popover
    </button>
  </div>
</div>
`;
